includes/class-wc-structured-data.php

```diff
@@ -0,0 +1,500 @@
+		add_action( 'woocommerce_shop_loop',              array( $this, 'generate_product_data' ), 10 );
+		add_action( 'woocommerce_single_product_summary', array( $this, 'generate_product_data' ), 60 );
+		add_action( 'woocommerce_review_meta',            array( $this, 'generate_product_review_data' ), 20, 1 );
+		add_action( 'woocommerce_email_order_details',    array( $this, 'generate_email_order_data' ), 20, 4 );
+		// Enqueue structured data...
+		add_action( 'woocommerce_email_order_details',    array( $this, 'enqueue_data' ), 30 );
+		add_action( 'wp_footer',                          array( $this, 'enqueue_data_type_for_page' ) );
+	}
+
+	/**
+	 * Sets `$this->_data`.
+	 *
+	 * @param  array $data
+	 * @param  bool $reset (default: false)
+	 * @return bool
+	 */
+	public function set_data( $data, $reset = false ) {
+		if ( ! is_array( $data ) || ! array_key_exists( '@type', $data ) ) {
+			return false;
+		}
+
+		if ( $reset && isset( $this->_data ) ) {
+			unset( $this->_data );
+		}
+		
+		$this->_data[] = $data;
+
+		return true;
+	}
+	
+	/**
+	 * Gets `$this->_data`.
+	 *
+	 * @return array $data
+	 */
+	public function get_data() {
+		return $data = isset( $this->_data ) ? $this->_data : array();
+	}
+
+	/**
+	 * Sets `$this->_structured_data`.
+	 *
+	 * @return bool
+	 */
+	public function set_structured_data() {
+		if ( isset( $this->_structured_data ) || ! $structured_data = $this->get_data() ) {
+			return false;
+		}
+
+		foreach ( $structured_data as $value ) {
+			switch ( $type = $value['@type'] ) {
+				case 'MusicAlbum':
+				case 'SoftwareApplication':
+					$type = 'Product';
+					break;
+			}
+
+			$structured_data[ $type ][] = $value;
+		}
+
+		foreach ( $structured_data as $type => $value ) {
+			if ( count( $value ) > 1 ) {
+				$structured_data[ $type ] = array( '@graph' => $value );
+			} else {
+				$structured_data[ $type ] = $value[0];
+			}
+
+			$structured_data[ $type ] = apply_filters( 'woocommerce_structured_data_context', array( '@context' => 'http://schema.org/' ), $structured_data, $type, $value ) + $structured_data[ $type ];
+		}
+
+		$this->_structured_data = $structured_data;
+
+		return true;
+	}
+
+	/**
+	 * Gets `$this->_structured_data`.
+	 * 
+	 * @return array $structured_data
+	 */
+	public function get_structured_data() {
+		$this->set_structured_data();
+
+		return $structured_data = isset( $this->_structured_data ) ? $this->_structured_data : array();
+	}
+
+	/**
+	 * Sanitizes, encodes and echoes structured data.
+	 * 
+	 * List of the types available by default for specific request:
+	 * 'Product',
+	 * 'Review',
+	 * 'BreadcrumbList',
+	 * 'WebSite',
+	 * 'Order',
+	 *
+	 * @uses   `woocommerce_email_order_details` action hook
+	 * @param  bool|array $requested_types (default: false)
+	 * @return bool
+	 */
+	public function enqueue_data( $requested_types = false ) {
+		if ( ! $structured_data = $this->get_structured_data() ) {
+			return false;
+		}
+
+		if ( $requested_types ) {
```

Comment:
I don't know if I'm missing something obvious or if I get a very odd bug here.

`add_action( 'woocommerce_email_order_details',    array( $this, 'enqueue_data' ), 30 );` will actually enter this condition... So it will fail to enqueue data in the email because it also enter the next condition below since $requested_types is not an array and should not even be anything else than `false` ! ... WTH??